keps/draft-repo.md

```diff
@@ -0,0 +1,108 @@
+
+### Goals
+
+- Automate the creation of a operator tarball
+- Automate the creation or updating of the repository index with a new operator tarball
+- Define rules for incubator and graduated along with graduation tooling.
+- Define how an admin can create their own repository
+- Install operator from incubator instead of graduated repository
+
+### Non-Goals
+
+- Manage repos of non-KUDO applications (e.g. Helm Charts)
+- External formats defined by [KEP0013](keps/0013-external-specs.md)
+- Repository cacheing
+
+## Proposal
+
+### Create Operator Tarball
+
+An operator package / bundle as defined in [KEP0010](keps/0010-package-manager.md).  The storage package at the repository is a tarball. We need a way to create this tarball in a uniformed way. KUDO shall a way to generate a tarball based on the standard file system layout.  As an example: `kubectl kudo package docs/examples/zookeeper/`.  This will create a `zookeeper-3.4.10.tgz` based on parsing the `operator.yaml` for version details. It would be best if this included linting to ensure that the static structure of the operator is correct. At this point, it will guarantee that operator folder has `operator.yaml` and `params.yaml` along with a templates directory.  It will create the following operator tarball `{operator.yaml:name}-{operator.yaml:version}.tgz`.
+
+### Create Repository Index
+
+KUDO needs the ability to create an index file for the repository. Something like `kubectl kudo repo index new-repo --url https://kudo-repo.storage.googleapis.com`. In this example, the `kudo repo index` is the command. The `new-repo` is a folder containing operator tarballs. The index file defined in [KEP0010](keps/0010-package-manager.md) will be created using the url provided for links in the file.
+
+Steps for creating an index file for a new operator looks like:
+```
+kubectl kudo package docs/examples/zookeeper/
+mkdir new-repo
+mv zookeeper-0.1.0.tgz new-repo/
+kubectl kudo repo index new-repo --url https://kudo-repo.storage.googleapis.com
+```
+
+### Create Local Respository
+
+For convenience it should be possible to have KUDO start a repository service with a provided path.  `kubectl kudo serve --repo-path ./repo` would create a local http service for the repo in `./repo`
+
+### Update a Repository
+
+Inspired by Helm, KUDO needs a way to take a repository index and merge it with another index for the purposes of adding operators to a repository.  Example: `kubectl kudo  repo --merge`.
+
+### CLI Repository Help
+
+`kubectl kudo  repo index --help`
+
+### User Stories
+
+- Allow pushing of new KUDO operator to the incubator repository
+- Allow running of a CNAB bundle as a KUDO Operator
+- Allow the running of an Operator as a KUDO Operator
+
+
+### Risks and Mitigations
+
+
+## Graduation Criteria
+
+We need to define the rules for promoting incubator operators to the graduated repository.
```

Comment:
I think at least one criteria would be some level of test coverage and passing in the Kubernetes environments defined in KEP-0004.